middle_end/flambda2/to_cmm/to_cmm_primitive.ml

```diff
@@ -557,193 +595,160 @@ let unary_float_arith_primitive _env dbg width op arg =
+          (Integral (Tagged _ | Untagged _) | Float (Float32 | Float64)) ) ->
+        None
+    in
+    extra, C.Scalar_type.static_cast ~dbg ~src ~dst arg
 
 let phys_equal _env dbg op x y =
   match (op : P.equality_comparison) with
   (* General case *)
   | Eq -> C.eq ~dbg x y
   | Neq -> C.neq ~dbg x y
 
+let requires_sign_extended_operands : P.binary_int_arith_op -> bool = function
+  | Div | Mod ->
+    (* Note that it would be wrong to apply [C.low_bits] to operands for div and
+       mod.
+
+       Some background: The problem arises in cases like: [(num1 * num2) /
+       num3]. If an overflow occurs in the multiplication, then we must deal
+       with it by sign-extending before the division. Whereas [ (num1 * num2) *
+       num3 ] can delay the sign-extension until the very end, even in the case
+       of overflow in the middle. So in a way, div and mod are regular
+       functions, while all the others are special as they can delay overflow
+       handling.
+
+       Cmm only has [Arch.size_int]-width virtual registers, so we must always
+       do operations on values of that size. (If we had smaller virtual
+       registers, we could use them in Cmm without sign-extension and let the
+       backend insert sign-extensions if it doesn't support operations on n-bit
+       physical registers. There was a prototype developed of this but it was
+       quite complicated and didn't get merged.) *)
+    true
+  | Add | Sub | Mul ->
+    (* https://en.wikipedia.org/wiki/Modular_arithmetic - these operations are
+       compatible with modular arithmetic *)
+    false
+  | And | Or | Xor ->
+    (* bitwise operations are clearly compatible *)
+    false
+
 let binary_int_arith_primitive _env dbg (kind : K.Standard_int.t)
     (op : P.binary_int_arith_op) x y =
-  match kind with
-  | Tagged_immediate -> (
-    match op with
-    | Add -> C.add_int_caml x y dbg
-    | Sub -> C.sub_int_caml x y dbg
-    | Mul -> C.mul_int_caml x y dbg
-    | Div -> C.div_int_caml x y dbg
-    | Mod -> C.mod_int_caml x y dbg
-    | And -> C.and_int_caml x y dbg
-    | Or -> C.or_int_caml x y dbg
-    | Xor -> C.xor_int_caml x y dbg)
-  | Naked_int32 -> (
-    (* Operations on 32-bit integer arguments must return something in the range
-       of 32-bit integers, hence the sign_extensions here. The [C.low_32]
-       operations (see above in [sign_extend_32_can_delay_overflow]) are used to
-       avoid unnecessary sign extensions, e.g. when chaining additions together.
-       Also see comment below about [C.low_32] in the [Div] and [Mod] cases. *)
-    let sign_extend_32_can_delay_overflow f =
-      C.sign_extend ~bits:32 ~dbg
-        (f (C.low_bits ~bits:32 ~dbg x) (C.low_bits ~bits:32 ~dbg y) dbg)
+  let kind = integral_of_standard_int kind in
+  let[@local] wrap f =
+    (* We cast the operands to the width that the operator expects, apply the
+       operator, and cast the result back. *)
```

Comment:
The code below is correct, but every time I read it, it takes a long time to understand that calls to `wrap` will always be called with a `f` that operate on the `operator_kind` defined below, and the comment is not helping much. So either the comment should clearly clarify this, or **much preferably** `wrap` should take the `operator_kind` as argument alongside `f`.